apps/server/src/modules/pseudonym/service/feathers-roster.service.ts

```diff
@@ -0,0 +1,243 @@
+	async getUsersMetadata(pseudonym: string): Promise<UserMetdata> {
+		const loadedPseudonym: Pseudonym = await this.findPseudonymByPseudonym(pseudonym);
+		const user: UserDO = await this.userService.findById(loadedPseudonym.userId);
+
+		const userMetadata: UserMetdata = {
+			data: {
+				user_id: user.id as string,
+				username: this.pseudonymService.getIframeSubject(loadedPseudonym.pseudonym),
+				type: this.getUserRole(user),
+			},
+		};
+
+		return userMetadata;
+	}
+
+	async getUserGroups(pseudonym: string, oauth2ClientId: string): Promise<UserGroups> {
+		const loadedPseudonym: Pseudonym = await this.findPseudonymByPseudonym(pseudonym);
+		const externalTool: ExternalTool = await this.validateAndGetExternalTool(oauth2ClientId);
+
+		let courses: Course[] = await this.getCoursesFromUsersPseudonym(loadedPseudonym);
+		courses = await this.filterCoursesByToolAvailability(courses, externalTool.id as string);
+
+		const userGroups: UserGroups = {
+			data: {
+				groups: courses.map((course) => {
+					return {
+						group_id: course.id,
+						name: course.name,
+						student_count: course.students.length,
+					};
+				}),
+			},
+		};
+
+		return userGroups;
+	}
+
+	async getGroup(courseId: EntityId, oauth2ClientId: string): Promise<Group> {
+		const course: Course = await this.courseService.findById(courseId);
+		const externalTool: ExternalTool = await this.validateAndGetExternalTool(oauth2ClientId);
+
+		await this.validateSchoolExternalTool(course.school.id, externalTool.id as string);
+		await this.validateContextExternalTools(courseId);
+
+		const [studentEntities, teacherEntities, substitutionTeacherEntities] = await Promise.all([
+			course.students.loadItems(),
+			course.teachers.loadItems(),
+			course.substitutionTeachers.loadItems(),
+		]);
+
+		const [students, teachers, substitutionTeachers] = await Promise.all([
+			Promise.all(studentEntities.map((user) => this.userService.findById(user.id))),
+			Promise.all(teacherEntities.map((user) => this.userService.findById(user.id))),
+			Promise.all(substitutionTeacherEntities.map((user) => this.userService.findById(user.id))),
+		]);
+
+		const [studentPseudonyms, teacherPseudonyms, substitutionTeacherPseudonyms] = await Promise.all([
+			this.getAndFilterPseudonyms(students, externalTool),
+			this.getAndFilterPseudonyms(teachers, externalTool),
+			this.getAndFilterPseudonyms(substitutionTeachers, externalTool),
+		]);
+
+		const allTeacherPseudonyms: Pseudonym[] = teacherPseudonyms.concat(substitutionTeacherPseudonyms);
+
+		const group: Group = {
+			data: {
+				students: studentPseudonyms.map((pseudonym: Pseudonym) => this.mapPseudonymToUserData(pseudonym)),
+				teachers: allTeacherPseudonyms.map((pseudonym: Pseudonym) => this.mapPseudonymToUserData(pseudonym)),
+			},
+		};
+
+		return group;
+	}
+
+	private async getAndFilterPseudonyms(users: UserDO[], externalTool: ExternalTool): Promise<Pseudonym[]> {
+		const pseudonyms: (Pseudonym | null)[] = await Promise.all(
```

Comment:
findOrCreatePseudonym does not return null, so you don't need to filter

Reply:
Totally right